src/main/java/org/xstefank/model/PersistentList.java

```diff
@@ -30,32 +33,146 @@ public PersistentList(String dirName, String fileName) {
+            fw.write("");
+        } catch (IOException e) {
+            log.error(e);
+            replaceFileWithEmptyOne();
+        }
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends String> c) {
+        if (containsAll(c) || !super.addAll(c)) {
+            return false;
+        }
         try (FileWriter fw = new FileWriter(file, true)) {
-            fw.write(username + "\n");
+            for (String element : c) {
+                fw.write(element + System.lineSeparator());
+            }
         } catch (IOException e) {
-            throw new IllegalStateException("Cannot write user to file", e);
+            log.error(e);
+            super.removeAll(c);
+            int index = size() - c.size() - 1;
+            if (index < 0) {
+                index = 0;
+            }
+            rewriteFileContent(index);
         }
+        return containsAll(c);
     }
 
-    public boolean hasUsername(String  username) {
-        return list.contains(username);
+    @Override
+    public boolean addAll(int index, Collection<? extends String> c) {
+        if (containsAll(c) || !super.addAll(index, c)) {
+            return false;
+        }
+        rewriteFileContent(index);
+        return containsAll(c);
     }
 
     private void loadFileContent(File file) {
         try (BufferedReader br = new BufferedReader(new FileReader(file))) {
             String line;
             while ((line = br.readLine()) != null) {
                 if (!line.startsWith("#")) {
-                    list.add(line);
+                    super.add(line);
                 }
             }
         } catch (IOException e) {
             throw new IllegalStateException("Cannot read file " + file.getPath(), e);
         }
     }
-}
+
+    private void rewriteFileContent(int fromIndex) {
+        String lastProcessedElement = null;
+        try (RandomAccessFile randomAccessFile = new RandomAccessFile(file.getAbsolutePath(), "rw")){
+            int characterCount = 0;
+            if (randomAccessFile.length() > 0 && fromIndex > 0) {
+                for (int i = 0; i < fromIndex; i++) {
+                    characterCount += get(i).length() + System.lineSeparator().length();
+                }
+            }
+            randomAccessFile.setLength(characterCount);
+            randomAccessFile.seek(characterCount);
+
+            for (int i = fromIndex; i < size(); i++) {
+                lastProcessedElement = get(i);
+                randomAccessFile.writeBytes(lastProcessedElement + System.lineSeparator());
+            }
+        } catch (IOException e) {
+            if (lastProcessedElement == null) {
+                throw new IllegalStateException(e);
+            }
+            log.errorf("Cannot write element \"%s\" to list, Stacktrace: %s", lastProcessedElement, e);
+            int index = indexOf(lastProcessedElement);
+            super.remove(index);
+            rewriteFileContent(index);
```

Comment:
I think it can't produce infinite loop.. Because I am removing every element that failed on previous line however, it could produce really long loop that will remove every single element from the list.

Reply:
@lukasburda I think you are removing element from parent list -- which is itself another error I missed so please take a look at that also -- index isn't changing as well as the content of the file so I still think that if some line is unable to be written you will cycle